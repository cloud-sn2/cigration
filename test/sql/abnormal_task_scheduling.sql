/** 
测试环境信息
CN：127.0.0.1
    WORKER1：worker_1_host/57637
    WORKER2：worker_2_host/57638
    WORKER3：worker_3_host/57639

worker_1_host、worker_2_host、worker_3_host：都是127.0.0.1
*/

set datestyle = ISO, YMD;
set search_path = cigration_regress_test, cigration, public;
set citus.shard_count = 8;
set client_min_messages to warning;

drop schema if exists cigration_regress_test cascade;
create schema cigration_regress_test;

--
-- 1. 建表
--

-- 创建1组亲和的分片表

create table dist1(c1 int primary key, c2 text);
create table dist2(c1 int primary key, c2 text);
create table dist3(c1 int primary key, c2 text);

select create_distributed_table('dist1','c1', colocate_with=>'none');
select create_distributed_table('dist2','c1', colocate_with=>'dist1');
select create_distributed_table('dist3','c1', colocate_with=>'dist1');

-- 插入数据
insert into dist1 select generate_series(1,100000), 'aaa';
insert into dist2 select generate_series(1,10000), 'bbb';
insert into dist3 select generate_series(1,100000), 'ccc';

-- 查看所有分片的初始分布
select nodename,
       nodeport,
       logicalrelid,
       shardminvalue,
       shardmaxvalue
from pg_dist_shard_placement p
     join pg_dist_shard s on(p.shardid = s.shardid)
where logicalrelid::text ~ 'dist'
order by nodename,nodeport,logicalrelid,shardminvalue;

--
-- 2. 创建迁移任务（init）
--

-- 创建迁移worker1的所有分片到worker3的job并获取第一个迁移任务的jobid和taskid
select jobid, taskid from cigration_create_move_node_job(:'worker_1_host', :worker_1_port, :'worker_3_host', :worker_3_port) limit 1 \gset

select source_nodename,source_nodeport,target_nodename,target_nodeport,status,total_shard_count
from pg_citus_shard_migration where jobid=:jobid and taskid=:taskid;

-- 迁移任务的初始状态为init
select status from pg_citus_shard_migration where jobid=:jobid and taskid=:taskid;
select cigration_monitor_shard_migration_task(:jobid, :taskid);

-- 禁止创建新的迁移作业
select jobid, taskid from cigration_create_drain_node_job(array[concat(:'worker_2_host', ':', :worker_2_port), concat(:'worker_3_host', ':', :worker_3_port)]);
select jobid, taskid from cigration_create_rebalance_job();
select jobid, taskid from cigration_create_move_node_job(:'worker_1_host', :worker_1_port, :'worker_3_host', :worker_3_port);

-- 禁止取消和完成init迁移任务
select cigration_cancel_shard_migration_task(:jobid, :taskid);
select cigration_complete_shard_migration_task(:jobid, :taskid);

-- 禁止对非completed迁移任务删除旧分片
select cigration_drop_old_shard(:jobid, :taskid);


--
-- 3. 启动迁移任务（running）
--
select cigration_start_shard_migration_task(:jobid, :taskid);
select status from pg_citus_shard_migration where jobid=:jobid and taskid=:taskid;

-- 禁止启动running迁移任务
select cigration_start_shard_migration_task(:jobid, :taskid);

-- 禁止对非completed迁移任务删除旧分片
select cigration_drop_old_shard(:jobid, :taskid);

-- 禁止清理running迁移任务
select cigration_cleanup_shard_migration_task(:jobid, :taskid);


--
-- 4. 取消迁移任务（canceled）
--

select cigration_cancel_shard_migration_task(:jobid, :taskid);
select status from pg_citus_shard_migration where jobid=:jobid and taskid=:taskid;
select cigration_monitor_shard_migration_task(:jobid, :taskid);

-- 禁止取消完成canceled迁移任务
select cigration_cancel_shard_migration_task(:jobid, :taskid);
select cigration_complete_shard_migration_task(:jobid, :taskid);


--
-- 5. 构造运行出错的迁移任务（init->error）
--

-- 注入故障
select target_nodename from pg_citus_shard_migration where jobid=:jobid and taskid=:taskid \gset
update pg_citus_shard_migration set target_nodename = 'faked_host' where jobid=:jobid and taskid=:taskid;

-- 启动迁移任务（error）
select cigration_start_shard_migration_task(:jobid, :taskid);
select status from pg_citus_shard_migration where jobid=:jobid and taskid=:taskid;
select cigration_monitor_shard_migration_task(:jobid, :taskid);

-- 禁止启动取消完成error迁移任务
select cigration_start_shard_migration_task(:jobid, :taskid);
select cigration_cancel_shard_migration_task(:jobid, :taskid);
select cigration_complete_shard_migration_task(:jobid, :taskid);

-- 禁止对非completed迁移任务删除旧分片
select cigration_drop_old_shard(:jobid, :taskid);

-- 禁止清理error迁移任务
select cigration_cleanup_shard_migration_task(:jobid, :taskid);

-- 解除故障
update pg_citus_shard_migration set target_nodename = :'target_nodename' where jobid=:jobid and taskid=:taskid;

-- 清理环境并恢复到init状态
select cigration_cleanup_error_env();
select status from pg_citus_shard_migration where jobid=:jobid and taskid=:taskid;


--
-- 6. 构造运行出错的迁移任务（running->running）
--

-- 启动迁移任务（running）
select cigration_start_shard_migration_task(:jobid, :taskid);
select status from pg_citus_shard_migration where jobid=:jobid and taskid=:taskid;

-- 注入故障
select target_nodename from pg_citus_shard_migration where jobid=:jobid and taskid=:taskid \gset
update pg_citus_shard_migration set target_nodename = 'faked_host' where jobid=:jobid and taskid=:taskid;

-- 完成迁移任务（running）
-- running状态下出错仍然是running
select cigration_complete_shard_migration_task(:jobid, :taskid);
select status from pg_citus_shard_migration where jobid=:jobid and taskid=:taskid;
select cigration_monitor_shard_migration_task(:jobid, :taskid);

-- 解除故障
update pg_citus_shard_migration set target_nodename = :'target_nodename' where jobid=:jobid and taskid=:taskid;


--
-- 7. 完成迁移任务（completed）
--

-- 检查迁移完成前，SQL执行正常
select count(*) from dist1 a left join dist2 b on(a.c1=b.c1) left join dist3 c on (a.c1=c.c1);

-- 完成迁移任务（completed）
select cigration_complete_shard_migration_task(:jobid, :taskid, init_sync_timeout=>10, data_sync_timeout=>20);
select status from pg_citus_shard_migration where jobid=:jobid and taskid=:taskid;
select cigration_monitor_shard_migration_task(:jobid, :taskid);

-- 检查迁移完成后，SQL执行正常
select count(*) from dist1 a left join dist2 b on(a.c1=b.c1) left join dist3 c on (a.c1=c.c1);

-- 禁止取消，启动，完成completed迁移任务
select cigration_cancel_shard_migration_task(:jobid, :taskid);
select cigration_complete_shard_migration_task(:jobid, :taskid);
select cigration_start_shard_migration_task(:jobid, :taskid);

--
-- 8. 归档迁移任务（history）
--
select cigration_cleanup_shard_migration_task(:jobid, :taskid);
select status from pg_citus_shard_migration where jobid=:jobid and taskid=:taskid;
select status from pg_citus_shard_migration_history where jobid=:jobid and taskid=:taskid;

-- 禁止对归档迁移任务执行调度操作
select cigration_cancel_shard_migration_task(:jobid, :taskid);
select cigration_complete_shard_migration_task(:jobid, :taskid);
select cigration_start_shard_migration_task(:jobid, :taskid);
select cigration_drop_old_shard(:jobid, :taskid);
select cigration_cleanup_shard_migration_task(:jobid, :taskid);


--
-- 9. 归档剩余迁移任务（history）
--

-- 设置一个任务为canceled状态，确认canceled任务可以被归档
select cigration_start_shard_migration_task(jobid, taskid)
from pg_citus_shard_migration order by taskid limit 1;

select cigration_cancel_shard_migration_task(jobid, taskid)
from pg_citus_shard_migration order by taskid limit 1;

-- 归档所有剩余迁移任务
select status from pg_citus_shard_migration order by taskid;

select cigration_cleanup_shard_migration_task(jobid, taskid) from pg_citus_shard_migration;

-- 归档所有剩余迁移任务到历史任务表
select count(*) from pg_citus_shard_migration;
select count(*) from pg_citus_shard_migration_history where jobid=:jobid;

--
-- 10. 旧分片清理
--

select nodename,nodeport,count(*)
from cigration_get_recyclebin_metadata()
group by nodename,nodeport
order by nodename,nodeport;

select cigration_cleanup_recyclebin();

select nodename,nodeport,count(*)
from cigration_get_recyclebin_metadata()
group by nodename,nodeport
order by nodename,nodeport;

