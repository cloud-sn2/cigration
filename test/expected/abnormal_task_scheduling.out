/** 
测试环境信息
CN：127.0.0.1
    WORKER1：worker_1_host/57637
    WORKER2：worker_2_host/57638
    WORKER3：worker_3_host/57639

worker_1_host、worker_2_host、worker_3_host：都是127.0.0.1
*/
set datestyle = ISO, YMD;
set search_path = cigration_regress_test, cigration, public;
set citus.shard_count = 8;
set client_min_messages to warning;
drop schema if exists cigration_regress_test cascade;
create schema cigration_regress_test;
--
-- 1. 建表
--
-- 创建1组亲和的分片表
create table dist1(c1 int primary key, c2 text);
create table dist2(c1 int primary key, c2 text);
create table dist3(c1 int primary key, c2 text);
select create_distributed_table('dist1','c1', colocate_with=>'none');
 create_distributed_table 
--------------------------
 
(1 row)

select create_distributed_table('dist2','c1', colocate_with=>'dist1');
 create_distributed_table 
--------------------------
 
(1 row)

select create_distributed_table('dist3','c1', colocate_with=>'dist1');
 create_distributed_table 
--------------------------
 
(1 row)

-- 插入数据
insert into dist1 select generate_series(1,100000), 'aaa';
insert into dist2 select generate_series(1,10000), 'bbb';
insert into dist3 select generate_series(1,100000), 'ccc';
-- 查看所有分片的初始分布
select nodename,
       nodeport,
       logicalrelid,
       shardminvalue,
       shardmaxvalue
from pg_dist_shard_placement p
     join pg_dist_shard s on(p.shardid = s.shardid)
where logicalrelid::text ~ 'dist'
order by nodename,nodeport,logicalrelid,shardminvalue;
 nodename  | nodeport | logicalrelid | shardminvalue | shardmaxvalue 
-----------+----------+--------------+---------------+---------------
 localhost |    57637 | dist1        | 1073741824    | 1610612735
 localhost |    57637 | dist1        | -2147483648   | -1610612737
 localhost |    57637 | dist1        | -536870912    | -1
 localhost |    57637 | dist2        | 1073741824    | 1610612735
 localhost |    57637 | dist2        | -2147483648   | -1610612737
 localhost |    57637 | dist2        | -536870912    | -1
 localhost |    57637 | dist3        | 1073741824    | 1610612735
 localhost |    57637 | dist3        | -2147483648   | -1610612737
 localhost |    57637 | dist3        | -536870912    | -1
 localhost |    57638 | dist1        | 0             | 536870911
 localhost |    57638 | dist1        | 1610612736    | 2147483647
 localhost |    57638 | dist1        | -1610612736   | -1073741825
 localhost |    57638 | dist2        | 0             | 536870911
 localhost |    57638 | dist2        | 1610612736    | 2147483647
 localhost |    57638 | dist2        | -1610612736   | -1073741825
 localhost |    57638 | dist3        | 0             | 536870911
 localhost |    57638 | dist3        | 1610612736    | 2147483647
 localhost |    57638 | dist3        | -1610612736   | -1073741825
 localhost |    57639 | dist1        | -1073741824   | -536870913
 localhost |    57639 | dist1        | 536870912     | 1073741823
 localhost |    57639 | dist2        | -1073741824   | -536870913
 localhost |    57639 | dist2        | 536870912     | 1073741823
 localhost |    57639 | dist3        | -1073741824   | -536870913
 localhost |    57639 | dist3        | 536870912     | 1073741823
(24 rows)

--
-- 2. 创建迁移任务（init）
--
-- 创建迁移worker1的所有分片到worker3的job并获取第一个迁移任务的jobid和taskid
select jobid, taskid from cigration_create_move_node_job(:'worker_1_host', :worker_1_port, :'worker_3_host', :worker_3_port) limit 1 \gset
select source_nodename,source_nodeport,target_nodename,target_nodeport,status,total_shard_count
from pg_citus_shard_migration where jobid=:jobid and taskid=:taskid;
 source_nodename | source_nodeport | target_nodename | target_nodeport | status | total_shard_count 
-----------------+-----------------+-----------------+-----------------+--------+-------------------
 localhost       |           57637 | localhost       |           57639 | init   |                 3
(1 row)

-- 迁移任务的初始状态为init
select status from pg_citus_shard_migration where jobid=:jobid and taskid=:taskid;
 status 
--------
 init
(1 row)

select cigration_monitor_shard_migration_task(:jobid, :taskid);
 cigration_monitor_shard_migration_task 
----------------------------------------
 task_is_not_running
(1 row)

-- 禁止创建新的迁移作业
select jobid, taskid from cigration_create_drain_node_job(array[concat(:'worker_2_host', ':', :worker_2_port), concat(:'worker_3_host', ':', :worker_3_port)]);
ERROR:  can not create a new job when there are some uncleanuped jobs.
CONTEXT:  PL/pgSQL function cigration_create_drain_node_job(text[]) line 16 at RAISE
select jobid, taskid from cigration_create_rebalance_job();
ERROR:  can not create a new job when there are some uncleanuped jobs.
CONTEXT:  PL/pgSQL function cigration_create_rebalance_job() line 16 at RAISE
select jobid, taskid from cigration_create_move_node_job(:'worker_1_host', :worker_1_port, :'worker_3_host', :worker_3_port);
ERROR:  can not create a new job when there are some uncleanuped jobs.
CONTEXT:  PL/pgSQL function cigration_create_move_node_job(text,integer,text,integer) line 17 at RAISE
-- 禁止取消和完成init迁移任务
select cigration_cancel_shard_migration_task(:jobid, :taskid);
ERROR:  only the task in running status can be canceled.
CONTEXT:  PL/pgSQL function cigration_cancel_shard_migration_task(integer,integer) line 35 at RAISE
select cigration_complete_shard_migration_task(:jobid, :taskid);
ERROR:  only running status is supported,but now is init.
CONTEXT:  PL/pgSQL function cigration_complete_shard_migration_task(integer,integer,integer,integer,bigint,integer,boolean) line 46 at RAISE
-- 禁止对非completed迁移任务删除旧分片
select cigration_drop_old_shard(:jobid, :taskid);
ERROR:  task [52] in job [6] is not completed.
CONTEXT:  PL/pgSQL function cigration_drop_old_shard(integer,integer) line 28 at RAISE
--
-- 3. 启动迁移任务（running）
--
select cigration_start_shard_migration_task(:jobid, :taskid);
 cigration_start_shard_migration_task 
--------------------------------------
 init_succeed
(1 row)

select status from pg_citus_shard_migration where jobid=:jobid and taskid=:taskid;
 status  
---------
 running
(1 row)

-- 禁止启动running迁移任务
select cigration_start_shard_migration_task(:jobid, :taskid);
ERROR:  jobid:[6] task:[52] does not exists or the status is not "init".
CONTEXT:  PL/pgSQL function cigration_start_shard_migration_task(integer,integer,interval,boolean) line 23 at RAISE
-- 禁止对非completed迁移任务删除旧分片
select cigration_drop_old_shard(:jobid, :taskid);
ERROR:  task [52] in job [6] is not completed.
CONTEXT:  PL/pgSQL function cigration_drop_old_shard(integer,integer) line 28 at RAISE
-- 禁止清理running迁移任务
select cigration_cleanup_shard_migration_task(:jobid, :taskid);
ERROR:  only the completed,canceled or init task can be archived.
HINT:  if the task status is error,call function cigration_cleanup_error_env first.
CONTEXT:  PL/pgSQL function cigration_cleanup_shard_migration_task(integer,integer) line 21 at RAISE
--
-- 4. 取消迁移任务（canceled）
--
select cigration_cancel_shard_migration_task(:jobid, :taskid);
 cigration_cancel_shard_migration_task 
---------------------------------------
 cancel_succeed
(1 row)

select status from pg_citus_shard_migration where jobid=:jobid and taskid=:taskid;
  status  
----------
 canceled
(1 row)

select cigration_monitor_shard_migration_task(:jobid, :taskid);
 cigration_monitor_shard_migration_task 
----------------------------------------
 task_is_not_running
(1 row)

-- 禁止取消完成canceled迁移任务
select cigration_cancel_shard_migration_task(:jobid, :taskid);
ERROR:  only the task in running status can be canceled.
CONTEXT:  PL/pgSQL function cigration_cancel_shard_migration_task(integer,integer) line 35 at RAISE
select cigration_complete_shard_migration_task(:jobid, :taskid);
ERROR:  only running status is supported,but now is canceled.
CONTEXT:  PL/pgSQL function cigration_complete_shard_migration_task(integer,integer,integer,integer,bigint,integer,boolean) line 46 at RAISE
--
-- 5. 构造运行出错的迁移任务（init->error）
--
-- 注入故障
select target_nodename from pg_citus_shard_migration where jobid=:jobid and taskid=:taskid \gset
update pg_citus_shard_migration set target_nodename = 'faked_host' where jobid=:jobid and taskid=:taskid;
-- 启动迁移任务（error）
select cigration_start_shard_migration_task(:jobid, :taskid);
ERROR:  could not establish connection
DETAIL:  could not translate host name "faked_host" to address: Name or service not known
CONTEXT:  SQL statement "SELECT port 
                                 FROM dblink(format('host=%s port=%s user=%s dbname=%s', target_node_name, target_node_port, current_user, current_database()),
                      $$SELECT setting FROM pg_settings WHERE name = 'port'$$) AS tb(port integer)"
PL/pgSQL function cigration_move_shard_placement(integer,integer,interval,boolean) line 93 at SQL statement
SQL statement "select cigration.cigration_move_shard_placement(jobid_input,taskid_input,longtime_tx_threshold,with_replica_identity_check)"
PL/pgSQL function cigration_start_shard_migration_task(integer,integer,interval,boolean) line 27 at SQL statement
select status from pg_citus_shard_migration where jobid=:jobid and taskid=:taskid;
 status 
--------
 error
(1 row)

select cigration_monitor_shard_migration_task(:jobid, :taskid);
 cigration_monitor_shard_migration_task 
----------------------------------------
 task_is_not_running
(1 row)

-- 禁止启动取消完成error迁移任务
select cigration_start_shard_migration_task(:jobid, :taskid);
ERROR:  jobid:[6] task:[52] does not exists or the status is not "init".
CONTEXT:  PL/pgSQL function cigration_start_shard_migration_task(integer,integer,interval,boolean) line 23 at RAISE
select cigration_cancel_shard_migration_task(:jobid, :taskid);
ERROR:  only the task in running status can be canceled.
CONTEXT:  PL/pgSQL function cigration_cancel_shard_migration_task(integer,integer) line 35 at RAISE
select cigration_complete_shard_migration_task(:jobid, :taskid);
ERROR:  only running status is supported,but now is error.
CONTEXT:  PL/pgSQL function cigration_complete_shard_migration_task(integer,integer,integer,integer,bigint,integer,boolean) line 46 at RAISE
-- 禁止对非completed迁移任务删除旧分片
select cigration_drop_old_shard(:jobid, :taskid);
ERROR:  task [52] in job [6] is not completed.
CONTEXT:  PL/pgSQL function cigration_drop_old_shard(integer,integer) line 28 at RAISE
-- 禁止清理error迁移任务
select cigration_cleanup_shard_migration_task(:jobid, :taskid);
ERROR:  only the completed,canceled or init task can be archived.
HINT:  if the task status is error,call function cigration_cleanup_error_env first.
CONTEXT:  PL/pgSQL function cigration_cleanup_shard_migration_task(integer,integer) line 21 at RAISE
-- 解除故障
update pg_citus_shard_migration set target_nodename = :'target_nodename' where jobid=:jobid and taskid=:taskid;
-- 清理环境并恢复到init状态
select cigration_cleanup_error_env();
 cigration_cleanup_error_env 
-----------------------------
 
(1 row)

select status from pg_citus_shard_migration where jobid=:jobid and taskid=:taskid;
 status 
--------
 init
(1 row)

--
-- 6. 构造运行出错的迁移任务（running->running）
--
-- 启动迁移任务（running）
select cigration_start_shard_migration_task(:jobid, :taskid);
 cigration_start_shard_migration_task 
--------------------------------------
 init_succeed
(1 row)

select status from pg_citus_shard_migration where jobid=:jobid and taskid=:taskid;
 status  
---------
 running
(1 row)

-- 注入故障
select target_nodename from pg_citus_shard_migration where jobid=:jobid and taskid=:taskid \gset
update pg_citus_shard_migration set target_nodename = 'faked_host' where jobid=:jobid and taskid=:taskid;
-- 完成迁移任务（running）
-- running状态下出错仍然是running
select cigration_complete_shard_migration_task(:jobid, :taskid);
ERROR:  could not establish connection
DETAIL:  could not translate host name "faked_host" to address: Name or service not known
CONTEXT:  SQL statement "SELECT dblink_connect(dblink_target_name,format('host=%s port=%s user=%s dbname=%s',target_node_name,i_target_node_port,current_user,current_database()))"
PL/pgSQL function cigration_complete_shard_migration_task(integer,integer,integer,integer,bigint,integer,boolean) line 73 at PERFORM
select status from pg_citus_shard_migration where jobid=:jobid and taskid=:taskid;
 status  
---------
 running
(1 row)

select cigration_monitor_shard_migration_task(:jobid, :taskid);
ERROR:  could not establish connection
DETAIL:  could not translate host name "faked_host" to address: Name or service not known
CONTEXT:  SQL statement "SELECT dblink_connect(dblink_target_name,format('host=%s port=%s user=%s dbname=%s',target_node_name,i_target_node_port,current_user,current_database()))"
PL/pgSQL function cigration_complete_shard_migration_task(integer,integer,integer,integer,bigint,integer,boolean) line 73 at PERFORM
SQL statement "select cigration.cigration_complete_shard_migration_task(jobid_input,taskid_input,check_only=>true)"
PL/pgSQL function cigration_monitor_shard_migration_task(integer,integer) line 5 at SQL statement
-- 解除故障
update pg_citus_shard_migration set target_nodename = :'target_nodename' where jobid=:jobid and taskid=:taskid;
--
-- 7. 完成迁移任务（completed）
--
-- 检查迁移完成前，SQL执行正常
select count(*) from dist1 a left join dist2 b on(a.c1=b.c1) left join dist3 c on (a.c1=c.c1);
 count  
--------
 100000
(1 row)

-- 完成迁移任务（completed）
select cigration_complete_shard_migration_task(:jobid, :taskid, init_sync_timeout=>10, data_sync_timeout=>20);
 cigration_complete_shard_migration_task 
-----------------------------------------
 complete
(1 row)

select status from pg_citus_shard_migration where jobid=:jobid and taskid=:taskid;
  status   
-----------
 completed
(1 row)

select cigration_monitor_shard_migration_task(:jobid, :taskid);
 cigration_monitor_shard_migration_task 
----------------------------------------
 task_is_not_running
(1 row)

-- 检查迁移完成后，SQL执行正常
select count(*) from dist1 a left join dist2 b on(a.c1=b.c1) left join dist3 c on (a.c1=c.c1);
 count  
--------
 100000
(1 row)

-- 禁止取消，启动，完成completed迁移任务
select cigration_cancel_shard_migration_task(:jobid, :taskid);
ERROR:  only the task in running status can be canceled.
CONTEXT:  PL/pgSQL function cigration_cancel_shard_migration_task(integer,integer) line 35 at RAISE
select cigration_complete_shard_migration_task(:jobid, :taskid);
ERROR:  only running status is supported,but now is completed.
CONTEXT:  PL/pgSQL function cigration_complete_shard_migration_task(integer,integer,integer,integer,bigint,integer,boolean) line 46 at RAISE
select cigration_start_shard_migration_task(:jobid, :taskid);
ERROR:  jobid:[6] task:[52] does not exists or the status is not "init".
CONTEXT:  PL/pgSQL function cigration_start_shard_migration_task(integer,integer,interval,boolean) line 23 at RAISE
--
-- 8. 归档迁移任务（history）
--
select cigration_cleanup_shard_migration_task(:jobid, :taskid);
 cigration_cleanup_shard_migration_task 
----------------------------------------
 
(1 row)

select status from pg_citus_shard_migration where jobid=:jobid and taskid=:taskid;
 status 
--------
(0 rows)

select status from pg_citus_shard_migration_history where jobid=:jobid and taskid=:taskid;
  status   
-----------
 completed
(1 row)

-- 禁止对归档迁移任务执行调度操作
select cigration_cancel_shard_migration_task(:jobid, :taskid);
ERROR:  only the task in running status can be canceled.
CONTEXT:  PL/pgSQL function cigration_cancel_shard_migration_task(integer,integer) line 35 at RAISE
select cigration_complete_shard_migration_task(:jobid, :taskid);
ERROR:  query returned no rows
CONTEXT:  PL/pgSQL function cigration_complete_shard_migration_task(integer,integer,integer,integer,bigint,integer,boolean) line 41 at SQL statement
select cigration_start_shard_migration_task(:jobid, :taskid);
ERROR:  jobid:[6] task:[52] does not exists or the status is not "init".
CONTEXT:  PL/pgSQL function cigration_start_shard_migration_task(integer,integer,interval,boolean) line 23 at RAISE
select cigration_drop_old_shard(:jobid, :taskid);
ERROR:  task [52] in job [6] is not completed.
CONTEXT:  PL/pgSQL function cigration_drop_old_shard(integer,integer) line 28 at RAISE
select cigration_cleanup_shard_migration_task(:jobid, :taskid);
ERROR:  only the completed,canceled or init task can be archived.
HINT:  if the task status is error,call function cigration_cleanup_error_env first.
CONTEXT:  PL/pgSQL function cigration_cleanup_shard_migration_task(integer,integer) line 21 at RAISE
--
-- 9. 归档剩余迁移任务（history）
--
-- 设置一个任务为canceled状态，确认canceled任务可以被归档
select cigration_start_shard_migration_task(jobid, taskid)
from pg_citus_shard_migration order by taskid limit 1;
 cigration_start_shard_migration_task 
--------------------------------------
 init_succeed
(1 row)

select cigration_cancel_shard_migration_task(jobid, taskid)
from pg_citus_shard_migration order by taskid limit 1;
 cigration_cancel_shard_migration_task 
---------------------------------------
 cancel_succeed
(1 row)

-- 归档所有剩余迁移任务
select status from pg_citus_shard_migration order by taskid;
  status  
----------
 canceled
 init
(2 rows)

select cigration_cleanup_shard_migration_task(jobid, taskid) from pg_citus_shard_migration;
 cigration_cleanup_shard_migration_task 
----------------------------------------
 
 
(2 rows)

-- 归档所有剩余迁移任务到历史任务表
select count(*) from pg_citus_shard_migration;
 count 
-------
     0
(1 row)

select count(*) from pg_citus_shard_migration_history where jobid=:jobid;
 count 
-------
     3
(1 row)

--
-- 10. 旧分片清理
--
select nodename,nodeport,count(*)
from cigration_get_recyclebin_metadata()
group by nodename,nodeport
order by nodename,nodeport;
 nodename  | nodeport | count 
-----------+----------+-------
 localhost |    57637 |     1
 localhost |    57638 |     1
 localhost |    57639 |     1
(3 rows)

select cigration_cleanup_recyclebin();
 cigration_cleanup_recyclebin 
------------------------------
 
(1 row)

select nodename,nodeport,count(*)
from cigration_get_recyclebin_metadata()
group by nodename,nodeport
order by nodename,nodeport;
 nodename | nodeport | count 
----------+----------+-------
(0 rows)

